Update to class based views

DrEdi opened the pull request:
All views now in class based style.

DrEdi committed:
Synchronize with master

DrEdi commented on main/templates/main/post_edit.html:
```diff
@@ -1,9 +1,8 @@
 {% extends 'main/base.html' %}
 
 {% block content %}
-    <h1>New post</h1>
```
I made it multipurpose for all classes that need form for their work.
So, I've deleted this string and change text in string a bit lower.
I think it's better to have one multipurpose than some only for one need

webknjaz replied:
Please read about YAGNI principle.

Miseracle commented on main/views.py:
```diff
@@ -1,102 +1,115 @@
+    def form_valid(self, form):
+        """Add info to form that were not given from POST request."""
+        form.instance.author = self.request.user
+        return super(NewPost, self).form_valid(form)
```
Mb you can use "super.form_valid(form)" syntax?

DrEdi replied:
Nope. `type object 'super' has no attribute 'form_valid'`

webknjaz replied:
`super()`

webknjaz replied:
@DrEdi in Python3 `super()` is a shortcut for `super(CurrentClass, self)`

webknjaz replied:
And I think you missed passing `self`

DrEdi replied:
If you mean to .form_valid(form) - it's ok.

webknjaz commented on blog/settings.py:
```diff
@@ -2,7 +2,7 @@
 import environ
 
-env = environ.Env(DEBUG=(bool, False),)
+env = environ.Env(DEBUG=(bool, True),)
```
Please revert this change. defaults are better to be production-ready. add `DEBUG=on` to `.env` instead.

DrEdi replied:
Could I do this when merging? In master branch it's ok. Just for minimize count of commits

webknjaz replied:
No, you have to react to comments instantly, otherwise you risk forgetting about them.

DrEdi replied:
Ok )

webknjaz commented on main/templates/main/index.html:
```diff
@@ -1,7 +1,7 @@
 
 {% block content %}
-    {% for post in posts %}
+    {% for post in object_list %}
```
It is possible to rename the queryset result objects name for template context in class-based views. Do this.

webknjaz commented on main/models.py:
```diff
@@ -33,7 +33,7 @@ def approved_comments(self):
     """Model for comments."""
 
-    post = models.ForeignKey('main.Post', related_name='comments')
+    post = models.ForeignKey('main.Post', related_name='comments', on_delete=models.CASCADE)
```
How does this change relate to the goal of the PR? (changing function-views with class-based ones)

DrEdi replied:
Sorry, I thought it was a good idea

webknjaz replied:
It may be a good idea if submitted separately

webknjaz replied:
or explained why it's necessary here/how it helps to achieve the goal of **current task**

DrEdi replied:
I understood, sorry for my fault

webknjaz commented on main/templates/main/post_draft_list.html:
```diff
@@ -1,7 +1,7 @@
 
 {% block content %}
-    {% for post in posts %}
+    {% for post in object_list %}
```
the above comment applies here as well

webknjaz commented on main/views.py:
```diff
@@ -1,102 +1,115 @@
+from django.shortcuts import get_object_or_404, redirect
+from django.core.urlresolvers import reverse_lazy
+from django.views.generic import DetailView, ListView, CreateView, UpdateView, View, DeleteView, TemplateView
+from .forms import CommentForm
```
I would separate project-imports from installed packages with an empty string.

webknjaz commented on main/views.py:
```diff
@@ -1,102 +1,115 @@
+    @method_decorator(login_required)
+    def dispatch(self, *args, **kwargs):
+        """Rewrite standart method and it to decorator."""
+        return super(Protected, self).dispatch(*args, **kwargs)
```
`super`'s arguments may be omitted in Python3 and you should pass `self` to dispatch.

DrEdi committed:
Fix remarks

Add human readable names for context data in views
Change debug mode
Separate project-imports from installed packages

webknjaz commented on main/views.py:
```diff
@@ -1,102 +1,118 @@
+    def form_valid(self, form):
+        """Add info to form that were not given from POST request."""
+        form.instance.author = self.request.user
+        return super().form_valid(form)
```
I think you missed `self.

webknjaz commented on main/views.py:
```diff
@@ -1,102 +1,118 @@
+
+    def form_valid(self, form):
+        """Add info to form that were not given from POST request."""
+        form.instance.author = self.request.user
```
I think it shouldn't go here.

webknjaz replied:
Try bypassing `request` object into `ModelForm`'s initializer like in https://github.com/open-craft-guild/crm-smsfly/blob/master/SMSFlyCRM/SMSApp/forms/task.py#L44-L46

You may set `form_class` instead of `model` in views :)

webknjaz commented on main/views.py:
```diff
@@ -1,102 +1,118 @@
+    @method_decorator(login_required)
+    def dispatch(self, *args, **kwargs):
+        """Rewrite standart method and it to decorator."""
+        return super().dispatch(self, *args, **kwargs)
```
It looks I was wrong and `self` is being passed automatically in this case

DrEdi replied:
Ok :+1:

DrEdi committed:
Fix class based views

DrEdi committed:
Fix style problem

webknjaz commented on vagga.yaml:
```diff
@@ -25,9 +25,9 @@ containers:
+      - !Py3Requirements requirements/production.txt
     environ:
-      DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/mainDB
+      DATABASE_URL: sqlite://vagga:vagga@127.0.0.1:5433/mainDB
```
Please don't use sqlite DB.

DrEdi replied:
This container will not run with postgres DB. We move all setting to other container. It's not need here anymore.

webknjaz replied:
WAT?
![](http://weknowmemes.com/wp-content/uploads/2013/09/wat-the-fuck.jpg)

DrEdi replied:
Omg, that was bad idea.
I reviewed code, just forget this...

webknjaz commented on vagga.yaml:
```diff
@@ -25,9 +25,9 @@ containers:
       - !Install
         - libpq
-      - !Py3Requirements requirements/main.txt
+      - !Py3Requirements requirements/production.txt
```
plz explain this change

DrEdi replied:
This container need requirements from production container. We run it with whitnoise.
This was my fault before

webknjaz replied:
okay

DrEdi commented on vagga.yaml:
```diff
@@ -25,9 +25,9 @@ containers:
+      - !Py3Requirements requirements/production.txt
     environ:
-      DATABASE_URL: postgresql://vagga:vagga@127.0.0.1:5433/mainDB
+      DATABASE_URL: sqlite://vagga:vagga@127.0.0.1:5433/mainDB
```
It's not working with postgres but works sqlite :D
I don't know why

DrEdi committed:
Fix DB link in vagga

webknjaz commented on main/views.py:
```diff
@@ -1,104 +1,136 @@
+    """Protect views that need to show only for authorised users."""
 
-def post_list(request):
+    @method_decorator(login_required)
```
Is it possible to apply `@login_required` to class?

DrEdi replied:
I created new class 'Protected' and inherits from this. 
Also I know about standart mixing class for log in, but unfortunately this one was not working for me :\

webknjaz replied:
what do you mean by inheriting `this`?

DrEdi replied:
from it*

webknjaz replied:
but the question was whether `@login_required` **can** be a class decorator

DrEdi replied:
We can use `@method_decorator` and there put `@login_required`. But IMHO this not very clear way

webknjaz replied:
```
@login_required
class BlahBla:
    pass
```

DrEdi replied:
This is impossible. `@login_required` is only for functions

webknjaz replied:
That's what I was questioning about

DrEdi replied:
Nope, we can't :)

webknjaz commented on main/views.py:
```diff
@@ -1,104 +1,136 @@
+    context_object_name = 'posts'
+    template_name = 'main/index.html'
+
+    def get_queryset(self):
```
AFAIR you could just assign `queryset` property.

webknjaz commented on main/views.py:
```diff
@@ -1,104 +1,136 @@
+        return super().form_valid(form)
+
+    def post(self, request, *args, **kwargs):
+        """Rewritten post method for saving the form."""
```
This isn't strictly valid description

webknjaz commented on main/views.py:
```diff
@@ -1,104 +1,136 @@
 
-@login_required
-def post_edit(request, pk):
+class EditPost(UpdateView, Protected):
```
Do you have tests for this?

DrEdi replied:
Yep, line 75-86 in test_views.py

webknjaz commented on main/views.py:
```diff
@@ -1,104 +1,136 @@
+
+    def post(self, request, *args, **kwargs):
+        """Add comment to DB."""
+        my_post = get_object_or_404(Post, pk=kwargs['pk'])
```
why `my_post`?

DrEdi replied:
Because 'post' would conflict with method name

webknjaz replied:
really? any proof? 
![](https://imgflip.com/s/meme/Computer-Guy-Facepalm.jpg)

DrEdi replied:
I thought :)

webknjaz commented on main/views.py:
```diff
@@ -1,104 +1,136 @@
+    fields = ['author', 'text']
+    template_name = 'main/post_edit.html'
+
+    def post(self, request, *args, **kwargs):
```
you could expose `pk` right here

DrEdi committed:
Rewtire queryset and docstring for method

DrEdi committed:
Revert to queryset method and change var name

webknjaz commented on main/views.py:
```diff
@@ -24,7 +24,10 @@ class PostList(ListView):
 
     context_object_name = 'posts'
     template_name = 'main/index.html'
-    queryset = Post.objects.filter(published_date__lte=timezone.now())
```
doesn't this work?

DrEdi replied:
Nope, here we, unfortunately, can't use it.
http://stackoverflow.com/questions/19707237/use-get-queryset-method-or-set-queryset-variable
